feat: browser support

OTArchive opened the pull request:
# Description

Support compiling to javascript and webassembly using Emscripten.

Compiling guide (PT-BR): [Guia ‐ OTClient Redemption Web](https://github.com/OTArchive/otclient-web/wiki/Guia-%E2%80%90-OTClient-Redemption-Web)

## Notes:

- Changed some .lua files encoding from UTF8-BOM to UTF8 (incompatible with browsers);
- Lua 5.1 is back (only when using emscripten), Luajit is incompatible with browsers;
- ~Changed .bit operations to a lua implementation (Lua 5.1 incompatible);~ Now using bitlib;
- Fixed some file names causing case-sensitivity issues (incompatible with browsers and unix systems);
- ~Currently only works with protocols that use http login servers.~ Now also supports older protocols.

IvanClementino committed:
browser support

mehah committed:
Merge branch 'main' into browser

rodrigopaixaorj committed:
Merge branch 'main' into browser

gesior commented:
@OTArchive 
It's the best merge into OTClient since OTCv8 source went public (otcv8-dev).
It should be merged as soon as possible (fix Builds).
We can drop LuaJIT and Lua 5.1+ features just to get client being able to run on website. **It can change Tibia and OTSes forever.**

IvanClementino committed:
CMake: WASM should be tested and else left for Linux as before

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

IvanClementino committed:
Remove conflicting vcpkg.json files

IvanClementino committed:
Use existing base64 encoder

IvanClementino committed:
Check for emscripten before includes and declarations

IvanClementino committed:
Improve CMakeLists

IvanClementino committed:
Fix case-sensitivity issues

divinity76 commented:
Great job!
do you actually have a functional test build?

Edit: damn yes seems so: https://demo.otarchive.com/

divinity76 commented:
The tibia protocol is a custom binary protocol not compatible with the socket APIs available in javascript (webSocket, XMLHttpRequest, fetch) ... do you have a plan for dealing with that? A websocket<->TCP proxy could work.. (but it will be able to steal all usernames/passwords of tibia 7.6 and older clients, possibly newer protocols too, soooo there's that)

OTArchive commented:
> The tibia protocol is a custom binary protocol not compatible with the socket APIs available in javascript (webSocket, XMLHttpRequest, fetch) ... do you have a plan for dealing with that? A websocket<->TCP proxy could work.. (but it will be able to steal all usernames/passwords of tibia 7.6 and older clients, possibly newer protocols too, soooo there's that)

It's already mentioned in [the guide](https://github.com/OTArchive/otclient-web/wiki/Guia-%E2%80%90-OTClient-Redemption-Web), but in Portuguese: yes, we can use proxies. I've setup my testing server using websockify.
Check out this video showing the demo and then the modular version connecting to my server:
https://www.youtube.com/watch?v=sGuYmr728eY

conde2 commented:
Amazing contribution, here is my short.review

Would be better to have the bit implementations to be C++ binds instead of this Lua class, and only fallback to those C++ implementations when not using LuaJit

divinity76 commented on modules/game_interface/widgets/statsbar.lua:
```diff
@@ -309,14 +309,13 @@ function StatsBar.reloadCurrentStatsBarQuickInfo_state(localPlayer, now, old)
     end
-
-    local bitsChanged = bit.bxor(now, old)
+    local bitsChanged = Bit.operation(now, old, Bit.OR)
```
this is wrong, i think you meant 
```lua
local bitsChanged = Bit.operation(now, old, Bit.XOR)
```
or
```lua
local bitsChanged = now & ~old;
```

divinity76 commented:
> Amazing contribution, here is my short.review
> 
> Would be better to have the bit implementations to be C++ binds instead of this Lua class, and only fallback to those class when not using LuaJit

why do you care? are you worried about performance? 

not really sure why we're using the bit library anyway tho, quoting https://www.lua.org/manual/5.3/manual.html#3.4.2
```
 Lua supports the following bitwise operators:
    &: bitwise AND
    |: bitwise OR
    ~: bitwise exclusive OR
    >>: right shift
    <<: left shift
    ~: unary bitwise NOT
```
seems lua support the standard bitwise operators,


so does that mean
```lua
return Bit.operation(g_window.getKeyboardModifiers(), KeyboardCtrlModifier, Bit.AND) ~= 0
```
is better than
```lua
return (g_window.getKeyboardModifiers() & KeyboardCtrlModifier) ~= 0;
```
or
```lua
local bitsChanged = Bit.operation(now, old, Bit.XOR)
```
is better than
```lua
local bitsChanged = now & ~old;
```
? is it easier to read? more maintainable?

conde2 commented:
Some operations are very limited or not possible to implement in LUA only.

Moreover the project can only use Lua 5.1 because os the env system that is not present in other Lua versions.

So for making the bit operations more reliable it should be implemented as a C++ to Lua bind

Commit:
Merge branch 'main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

IvanClementino committed:
Implement bitlib and restore original .lua files bit operations

OTArchive commented:
@conde2 @divinity76 
- I've implemented bitlib so we have the same bit operations as Luajit;
- Removed a few includes from the original bitlib code so I've also added the lib files in the commit;
- Restored .lua files to their original usage of bit operations.

Should we also include Lua 5.1 code and precompiled lib in the repo, as it is not available in vcpkg?

conde2 commented:
That's look a good solution, we also avoid new contributors to be confused why to not use default lib functions.

About including Lua 5.1 what other options we have? Nonetheless I will leave this decision to @mehah and others contributors.

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

IvanClementino committed:
Support older protocols and improvements

OTArchive commented:
Now with support for older protocols!

https://github.com/user-attachments/assets/4a27b0d2-a9b0-4816-aab0-f048102e006c

OTArchive committed:
Merge branch 'mehah:main' into browser

IvanClementino committed:
Set correct numpad keys

IvanClementino committed:
Intercept all browser keys

IvanClementino committed:
Allow pasting and writing symbols with shift

OTArchive committed:
Merge branch 'mehah:main' into browser

IvanClementino committed:
Improve networking
No need to reinvent the wheel

IvanClementino committed:
Reload page on client exit

IvanClementino committed:
Reduce CPU usage and connection improvements
Drop g_ioService to avoid doing work during polling and make the thread sleep a bit if no message ready

IvanClementino committed:
Disable Emscripten's text decoder
https://github.com/emscripten-core/emscripten/issues/18034

OTArchive committed:
Merge branch 'mehah:main' into browser

IvanClementino committed:
Basic mobile support
- Handle touch events and longpress for right click
- Check user agent to determine isMobile
- Use VirtualKeyboardAPI to show keyboard (iOS not supported https://developer.mozilla.org/en-US/docs/Web/API/VirtualKeyboard_API)

Commit:
Merge branch 'main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

IvanClementino committed:
Improve performance
Drop manually swapping buffers to reduce CPU overhead

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

IvanClementino committed:
Case insensitive FS and properly return if websocket fails to be created

IvanClementino committed:
Fix Vorbis Linking (7cd3c82)

IvanClementino committed:
Allow gameWorld port to be set without rebuilding
As servers and login servers are not adapted to send correct information when using the web client, it's necessary to override the world port and ip address that is received.
Before, the world port was being forced to what was defined in the WEBPORT variable, not matter what port was informed when connect() was called. Now, you can override the port using lua (characterlist.lua) and avoid rebuilding the client.
OBS: Port 7172 will still be overriden to 443 by the client.

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive commented on browser/overlay-ports/protobuf/fix-arm64-msvc.patch:
```diff
@@ -0,0 +1,22 @@
++#if defined(__aarch64__) && defined(PROTOBUF_LITTLE_ENDIAN) && !defined(_MSC_VER)
+   // This optimization is not supported in big endian mode
+   uint64_t first8;
+   std::memcpy(&first8, p, sizeof(first8));
```
Not sure! This patch was not made by me, it's one of the included with vcpkg. Mine is just the pthreads one.
Here is the related commit: https://github.com/microsoft/vcpkg/pull/40173

Can probably be removed from our repo, though untested

divinity76 commented on src/client/protocolgame.cpp:
```diff
@@ -33,7 +33,13 @@ void ProtocolGame::login(const std::string_view accountName, const std::string_v
+#ifndef __EMSCRIPTEN__
     connect(host, port);
+#else
+    if (port == 7172)
```
what's magic about 7172?

OTArchive replied:
Being the default game world port, it means that the user trying to use the web client didn't properly enforce the correct port to be used with websockets, so it's redirected to 443 for ease of use.

divinity76 commented on src/framework/const.h:
```diff
@@ -38,6 +38,11 @@
 #define BUILD_ARCH "ARM"
+#elif defined(__EMSCRIPTEN__)
+#ifndef WEBPORT
+#define WEBPORT 7979
```
is this used anywhere?

OTArchive replied:
> is this used anywhere?

Not anymore! Thanks for catching it

IvanClementino committed:
Remove unused definition

divinity76 commented on src/framework/core/graphicalapplication.cpp:
```diff
@@ -124,6 +128,37 @@ void GraphicalApplication::terminate()
+        return;
+    }
+
+    const auto& FPS = [this] {
```
nitpick remove the `&` , not enough of a c++ language lawyer to explain exactly why, but it is wrong.

ChatGPT-4o explains it as 
> Assigning a lambda to a const auto& can lead to dangling references because lambdas are temporary objects unless stored. ***This could cause undefined behavior.***

(That explanation doesn't sound quite right to me either but w/e)

OTArchive replied:
> nitpick remove the `&` , not enough of a c++ language lawyer to explain exactly why, but it is wrong.
> 
> ChatGPT-4o explains it as
> 
> > Assigning a lambda to a const auto& can lead to dangling references because lambdas are temporary objects unless stored. This could cause undefined behavior.
> 
> (That explanation doesn't sound quite right to me either but w/e)

Here you will need to help me a little bit please, because I copied it from line 178, without changes.

Edit: seem to work fine after removal, should it be removed from the original code too?

divinity76 replied:
>should it be removed from the original code too?

Yes.

IvanClementino committed:
Remove '&'
https://github.com/mehah/otclient/pull/894#discussion_r1807004213

divinity76 commented on src/framework/core/graphicalapplication.cpp:
```diff
@@ -124,6 +128,37 @@ void GraphicalApplication::terminate()
+    // update screen pixels
+    g_window.swapBuffers();
+
+    if (m_graphicFrameCounter.update()) {
+        g_dispatcher.addEvent([this, fps = FPS()] {
+            g_lua.callGlobalField("g_app", "onFps", fps);
+        });
+    }
```
```suggestion
    if (m_graphicFrameCounter.update()) {
        m_mapProcessFrameCounter.setTargetFps(
            (g_window.vsyncEnabled() || getMaxFps() || getTargetFps()) ? 500u : 0u
        );
        g_dispatcher.addEvent([this, fps = m_graphicFrameCounter.getFps()] {
            g_lua.callGlobalField("g_app", "onFps", fps);
        });
    }
```
then you can remove the entire FPS lambda 
(I'm sure the FPS lambda made sense in an earlier revision, but not anymore it seems)

mehah replied:
targetFPS needs to be set before update(), if you look inside update() method, the calculation is based on targetFPS.

https://github.com/mehah/otclient/blob/8d2c0651dd9cefa96728d7cfd2c95a5108dcafd8/src/framework/core/adaptativeframecounter.cpp#L30

IvanClementino committed:
Remove swapBuffers call, we no longer use it
https://github.com/mehah/otclient/commit/3c6bc5a5275ee421ebaf849e1c947fa4a2820668

divinity76 commented on src/framework/net/httplogin.cpp:
```diff
@@ -136,6 +142,63 @@ void LoginHttp::httpLogin(const std::string& host, const std::string& path,
+        };
+        attr.requestHeaders = headers;
+        attr.attributes = EMSCRIPTEN_FETCH_LOAD_TO_MEMORY | EMSCRIPTEN_FETCH_SYNCHRONOUS;
+        std::string body = "{\"email\":\"" + email + "\",\"password\":\"" + password + "\",\"stayloggedin\":true,\"type\":\"login\"}";
```
this only works as long as the password does not include any of `\` or `/` or `"`  or `'`

so technically this is wrong, but to fix it "properly" we need to introduce a proper JSON api/library,
 but as a quickfix we can do something like
```c++
std::string jsonEncodeString(const std::string_view input) {
    std::ostringstream encoded;
    
    for (char c : input) {
        switch (c) {
            case '"':  encoded << "\\\""; break;   // Escape double quotes
            case '\\': encoded << "\\\\"; break;   // Escape backslash
            case '\b': encoded << "\\b"; break;    // Escape backspace
            case '\f': encoded << "\\f"; break;    // Escape form feed
            case '\n': encoded << "\\n"; break;    // Escape new line
            case '\r': encoded << "\\r"; break;    // Escape carriage return
            case '\t': encoded << "\\t"; break;    // Escape tab
            default:
                // Check if the character is a control character or non-printable ASCII
                if (static_cast<unsigned char>(c) < 0x20 || static_cast<unsigned char>(c) > 0x7F) {
                    // Encode using Unicode escape
                    encoded << "\\u"
                            << std::hex << std::uppercase
                            << static_cast<int>(static_cast<unsigned char>(c) >> 4)
                            << static_cast<int>(static_cast<unsigned char>(c) & 0xF);
                } else {
                    encoded << c;  // Leave printable characters unchanged
                }
                break;
        }
    }
    
    return encoded.str();
}
```
and
```c++
        std::string body = "{\"email\":" + jsonEncodeString(email) + ",\"password\":" + jsonEncodeString(password) + ",\"stayloggedin\":true,\"type\":\"login\"}";
```

OTArchive replied:
On line 78 there is the original implementation using:
`json body = json{ {"email", email}, {"password", password}, {"stayloggedin", true}, {"type", "login"} };`

Should I use it instead?

divinity76 replied:
@OTArchive great, yes go for it 👍

divinity76 commented on src/framework/net/httplogin.cpp:
```diff
@@ -136,6 +142,63 @@ void LoginHttp::httpLogin(const std::string& host, const std::string& path,
+        attr.attributes = EMSCRIPTEN_FETCH_LOAD_TO_MEMORY | EMSCRIPTEN_FETCH_SYNCHRONOUS;
+        std::string body = "{\"email\":\"" + email + "\",\"password\":\"" + password + "\",\"stayloggedin\":true,\"type\":\"login\"}";
+        attr.requestData = body.c_str();
+        attr.requestDataSize = strlen(attr.requestData);
```
```suggestion
        attr.requestData = body.data();
        attr.requestDataSize = body.length();
```
in this case it doesn't really matter because we're not dealing with binary data, but if you did the c_str()+strlen approach and the body contained binary data, there would be data corruption.

IvanClementino committed:
Properly encode json
https://github.com/mehah/otclient/pull/894#discussion_r1807029895

divinity76 commented on src/framework/net/webconnection.cpp:
```diff
@@ -0,0 +1,290 @@
+    m_connecting = true;
+    m_connectCallback = connectCallback;
+
+    std::string ip = host.data();
```
```suggestion
    std::string ip = std::string(host);
```
(nitpick: .data() will perform a strlen()-style copy, doesn't really make sense, and its a little bit slower too, this will perform a memcpy()-style copy.)
Edit: sorry my original was a syntax error.

OTArchive replied:
I'm getting:
`error: no viable conversion from 'const std::string_view' (aka 'const basic_string_view<char>') to 'std::string' (aka 'basic_string<char>')`
Btw, thank you for explaining this stuff!

divinity76 replied:
> I'm getting: `error: no viable conversion from 'const std::string_view' (aka 'const basic_string_view<char>') to 'std::string' (aka 'basic_string<char>')` Btw, thank you for explaining this stuff!

sorry about that, fixed it now, re-read

divinity76 replied:
Could also mention when dealing with binary data
```
std::string copy = view.data();
```
will actually corrupt it (it will stop the copy at the first null byte), while
```
std::string copy = std::string(view);
```
will work with binary data too~

divinity76 commented on src/framework/net/webconnection.cpp:
```diff
@@ -0,0 +1,290 @@
+    const std::string prefix = "ws://";
+#else
+    const std::string prefix = "wss://";
+#endif
```
this looks odd, can you clarify? you're using un-encrypted websocket on release build, but using encrypted socket on debug build?

OTArchive replied:
It confused me at first too, but NDEBUG is defined when the build is set as release. It's un-encrypted when debugging.

OTArchive committed:
Update src/framework/net/webconnection.cpp

Co-authored-by: divinity76 <divinity76@gmail.com>

divinity76 commented on src/framework/net/webconnection.cpp:
```diff
@@ -0,0 +1,290 @@
+    const std::string prefix = "ws://";
+#else
+    const std::string prefix = "wss://";
+#endif
```
```suggestion
#ifdef NDEBUG
    const std::string prefix = "wss://";
#else
    const std::string prefix = "ws://";
#endif
```
the double negation thing can be confusing 😅  this is easier to read IMO

OTArchive committed:
Update src/framework/net/webconnection.cpp

Co-authored-by: divinity76 <divinity76@gmail.com>

divinity76 commented on src/framework/net/webconnection.cpp:
```diff
@@ -0,0 +1,290 @@
+    }));
+
+    emscripten_websocket_set_onmessage_callback(m_websocket, this, ([](int /*eventType*/, const EmscriptenWebSocketMessageEvent* webSocketEvent, void* userData) -> EM_BOOL {
+        uint32_t numBytes = webSocketEvent->numBytes;
```
```suggestion
        auto numBytes = webSocketEvent->numBytes;
```
it's actually uint64_t: https://github.com/emscripten-core/emscripten/blob/10cb9d46cdd17e7a96de68137c9649d9a630fbc7/system/include/emscripten/fetch.h#L164C1-L165C1
but easier to just let the compiler deduce it with `auto`

OTArchive committed:
Update src/framework/net/webconnection.cpp

Co-authored-by: divinity76 <divinity76@gmail.com>

divinity76 commented on src/framework/platform/browserplatform.cpp:
```diff
@@ -0,0 +1,133 @@
+
+void Platform::processArgs(std::vector<std::string>& /*args*/)
+{
+    //nothing todo, linux args are already utf8 encoded
```
That is not strictly true. Linux arguments does not have to be utf8, there is only 1 hard rule for linux shell arguments: they cannot contain null bytes (and they have a max length). Little test to prove it:
```php
$ cat utf8lol.php
<?php
$invalidUtf8 = "\xC2\x41";
$cmd = 'php -r "var_dump(mb_check_encoding(\$argv[1], \'UTF-8\'), \$argv[1]);" \'' . $invalidUtf8 . '\'';
system($cmd);
?>
$ php utf8lol.php
bool(false)
string(2) "�A"
```
- voila: passed a linux shell argument that was not valid UTF-8 ^^
- with the exception of null-bytes, linux shell arguments are actually binary safe ([when properly encoded](https://github.com/divinity76/phpquoteshellarg/blob/main/src/Divinity76/quoteshellarg/quoteshellarg.php#L37))
- i'm actually the author of a php library to properly encode linux shell arguments, because [PHP's escapeshellarg() does it wrong.](https://3v4l.org/Hkv7h)
But it is a very edge-case, no need to waste time on it here really.

divinity76 commented on src/framework/platform/browserplatform.cpp:
```diff
@@ -0,0 +1,133 @@
+        res += "/";
+    }
+    return res;
+}
```
we have c++17 yes?
```suggestion
std::string Platform::getCurrentDir()
{
    return std::filesystem::current_path();
}
```

divinity76 commented on src/framework/platform/browserplatform.cpp:
```diff
@@ -0,0 +1,133 @@
+std::string Platform::getTempPath()
+{
+    return "/tmp/";
+}
```
c++17
```suggestion
std::string Platform::getTempPath()
{
    return std::filesystem::temp_directory_path();
}
```

OTArchive committed:
Update src/framework/platform/browserplatform.cpp

Co-authored-by: divinity76 <divinity76@gmail.com>

OTArchive committed:
Update src/framework/platform/browserplatform.cpp

Co-authored-by: divinity76 <divinity76@gmail.com>

divinity76 commented on src/framework/platform/browserplatform.cpp:
```diff
@@ -0,0 +1,133 @@
+bool Platform::copyFile(std::string from, std::string to)
+{
+    return system(stdext::format("/bin/cp '%s' '%s'", from, to).c_str()) == 0;
+}
```
c++17
```suggestion
bool Platform::copyFile(std::string from, std::string to)
{
    return std::filesystem::copy_file(from, to);
}
```
(also there was technically a shell injection "vulnerability" in the original but w/e)

divinity76 commented on src/framework/platform/browserplatform.cpp:
```diff
@@ -0,0 +1,133 @@
+
+bool Platform::fileExists(std::string file)
+{
+    return true;
```
```suggestion
    return std::filesystem::exists(file);
```

OTArchive committed:
Update src/framework/platform/browserplatform.cpp

Co-authored-by: divinity76 <divinity76@gmail.com>

OTArchive committed:
Update src/framework/platform/browserplatform.cpp

Co-authored-by: divinity76 <divinity76@gmail.com>

divinity76 commented on src/framework/platform/browserplatform.cpp:
```diff
@@ -0,0 +1,133 @@
+bool Platform::removeFile(std::string file)
+{
+    return false;
+}
```
```suggestion
bool Platform::removeFile(std::string file, bool recursive = false)
{
    return (recursive ? std::filesystem::remove_all(file) : std::filesystem::remove(file));
}
```

divinity76 commented on src/framework/platform/browserplatform.cpp:
```diff
@@ -0,0 +1,133 @@
+
+ticks_t Platform::getFileModificationTime(std::string file)
+{
+    return 0;
```
c++17 has a lot of this stuff  :) but i'mma skip the rest of these for now

IvanClementino committed:
c++17 remove file

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

divinity76 commented on src/framework/platform/browserwindow.cpp:
```diff
@@ -0,0 +1,584 @@
+        }, this);
+        MAIN_THREAD_ASYNC_EM_ASM({
+            if (navigator && "virtualKeyboard" in navigator && (/iphone|ipod|ipad|android/i).test(navigator.userAgent)) {
+                navigator.virtualKeyboard.overlaysContent = true;
```
```suggestion
                const DOM_VK_BACK_SPACE = 8;
                const DOM_VK_RETURN = 13;
                const DOM_MYSTERY_CODE = 229;
                navigator.virtualKeyboard.overlaysContent = true;
```

divinity76 commented on src/framework/platform/browserwindow.cpp:
```diff
@@ -0,0 +1,584 @@
+    });
+}
+
+int number_of_characters_in_utf8_string(const char* str) {
```
nitpick: use size_t in case the string is over 2/4GB 😆

divinity76 commented on src/framework/platform/browserwindow.cpp:
```diff
@@ -0,0 +1,584 @@
+        b++;
+    }
+    return *a == *b;
+}
```
```suggestion
bool  string_eq(const char* a, const char* b) {
    if (a == nullptr || b == nullptr) return false;
    return std::strcmp(a, b) == 0;
}
```

divinity76 commented on src/framework/platform/browserwindow.cpp:
```diff
@@ -0,0 +1,584 @@
+    }
+    if (!keyboardEvent->repeat || eventType == EMSCRIPTEN_EVENT_KEYDOWN) {
+        Fw::Key keyCode = Fw::KeyUnknown;
+        for (int32_t i = 0; i < web_keymap.size(); i++) {
```
nit: size_t

IvanClementino committed:
Recommended improvements
https://github.com/mehah/otclient/pull/894#discussion_r1807112983
https://github.com/mehah/otclient/pull/894#discussion_r1807114089
https://github.com/mehah/otclient/pull/894#discussion_r1807115005
https://github.com/mehah/otclient/pull/894#discussion_r1807115609

divinity76 commented:
I have reviewed the code, and it looks good!  Awesome job! ❤️ 

There are stuff that can be improved (obviously), but I didn't see anything in the code that needs to be fixed before merging 

things that can be worked on: 
- would be nice if wss (encrypted websocket) support was configurable or auto-detected
- the port code looks a bit dodgy, not sure it would work with custom/non-standard ports.


But! have anyone looked into the failing Ubuntu builds? is it already broken or does this branch break it, or does this branch need to sync up with main? anyone knows?

OTArchive commented:
> Great job! I have reviewed the code, and it looks good! Awesome job! ❤️
> 
> There are stuff that can be improved (obviously), but I didn't see anything in the code that needs to be fixed before merging
> 
> things that can be worked on:
> 
>     * would be nice if wss (encrypted websocket) support was configurable or auto-detected
> 
>     * the port code looks a bit dodgy, not sure it would work with custom/non-standard ports.
> 
> 
> But! have anyone looked into the failing Ubuntu builds? is it already broken or does this branch break it, or does this branch need to sync up with main? anyone knows?

Thank you very much for the review @divinity76!

About wss, due to browser enforcement, we must use it when deployed due to a combination of headers and features that require https. Only when using localhost (not even 127.0.0.1) we will be able to use the client with ws.

About the port code, it's working fine with custom ports. Currently servers won't send the correct information to use websockets so we must enforce it (g_game.loginWorld in characterslist.lua).

Finally, about compiling, it just broke because of the latest commits on main. I'm sure @mehah will fix it soon.

divinity76 commented on src/framework/platform/browserplatform.cpp:
```diff
@@ -0,0 +1,127 @@
+
+ticks_t Platform::getFileModificationTime(std::string file)
+{
+    return 0;
```
```suggestion
    return std::filesystem::last_write_time(file);
```

divinity76 commented:
oh also, somewhere to configure the websockify proxy server to use will be needed at some point :)

OTArchive committed:
Merge branch 'mehah:main' into browser

divinity76 commented:
Just another nitpick, but it's something all C/C++ developers should use a little time to think about wrt coding style IMO,

While the choice between
```c
char* str;
```
and
```c
char *str;
```
is technically  a personal preference thing, the latter is *objectively better*.

Take a look at this bugged code:
```c
char* str1, str2;
```
Do you see the problem? It’s easy to miss. Now, let’s try it with the alternative style:
```c
char *str1, str2;
```
See it now? It's easier to see the bug with style 2.
They  `declare str1 as pointer to char, and declare str2 as a single char...` , when it was meant to be `declare str1 as pointer to char, and declare str2 as a pointer to char`

Fixing it, following coding style 1, it gets written as
```c
char* str1, *str2;
```
that looks... strange, doesn't it? compare that to style 2,
```c
char *str1, *str2;
```
looks much better imo,

and that is why, imo, we should all default to the style `char *str` instead of `char* str`  :)

ChatGPT can probably explain it better than me:
>(...) The * being next to the variable name emphasizes that the pointer is an attribute of the variable, not the char type itself. (...) it's immediately clear that the * applies to the variable, not the type. (...) So overall, I favor the char *str style for the same reasons you mentioned: readability, reducing confusion, and making the code less error-prone

(just a nitpick)

IvanClementino committed:
Merge branch 'mehah/main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

gesior commented:
> Just another nitpick, but it's something all C/C++ developers should use a little time to think about wrt coding style IMO,

**Is this problem related to this MR or whole repository?**

Can it be configured by `.editorconfig` (`https://github.com/mehah/otclient/blob/main/.editorconfig`) file somehow?

In PHP there is well known code-standard PSR-12 ( https://www.php-fig.org/psr/psr-12/ ), which replaces old standards PSR-1 ( https://www.php-fig.org/psr/psr-1/ ) and PSR-2 ( https://www.php-fig.org/psr/psr-2/ ).
PSR-12 fixes all possible problems (without forcing to reformat most of old code), so every PHP programmer uses it.

I've tried multiple C++ 'standards' in CLion:
![image](https://github.com/user-attachments/assets/30ed4ebc-ecc3-4056-b1a3-383973e9252e)
but all of them forces something different than TFS code format and all these formats looks weird.

FYI default `CLion` format moves all `*` to attributes/parameters variables, not their types - same as you posted.


Anytime I post anything to 2020+ PHP project, I can press CTRL+ALT+L - auto formatting - in PHPStorm to make my PHP code compatible with project (PSR-x).
For every OTS/OTC C++ project. I have to mark just few lines I've edited. Format them and then fix formatting manually in CLion.

OTArchive committed:
Merge branch 'mehah:main' into browser

IvanClementino committed:
Remove uneeded link libs

IvanClementino committed:
remove regex

OTArchive committed:
Merge branch 'mehah:main' into browser

IvanClementino committed:
Cleanup shell.html

IvanClementino committed:
Make compiling easier
- We can now use vcpkg.json with an adapted cmake command
(check guide)
- Update overlay-ports to the current baseline

IvanClementino committed:
Add baseline to overlay-ports vcpkg.json files

IvanClementino committed:
Workflows: ignore overlay-ports vcpkg.json

OTArchive commented:
Compiling is now streamlined and also uses vcpkg.json, so no need to install dependencies with separate vcpkg commands. A proper cmake command is all that is needed.
[Guide updated](https://github.com/OTArchive/otclient/wiki/Guia-%E2%80%90-OTClient-Redemption-Web).

OTArchive committed:
Merge branch 'mehah:main' into browser

IvanClementino committed:
Bot V8 basic fixes
Adapt load() calls to lua 5.1

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

OTArchive committed:
Merge branch 'mehah:main' into browser

gesior commented:
I think we can't keep this as PR forever. @OTArchive someday will get bored merging master over and over again and leave this branch unfinished.

If OTCR maintainers are not sure, if current version is stable enough/code is clean enough/build process is good enough (what about Lua 5.1/5.4/LuaJIT 2.1 compatibility on desktop version?), **then maybe we can merge this branch into official branch** `browser` **or** `webclient`?

**Maybe then more people will test it, merge master, refactor code, simplify/document configuration.**

OTArchive committed:
Merge branch 'mehah:main' into browser

mehah committed:
Merge branch 'main' into browser
